src/sentry/digests/notifications.py

```diff
@@ -91,41 +100,43 @@ def attach_state(project, groups, rules, event_counts, user_counts):
 
 
 class Pipeline:
-    def __init__(self):
-        self.operations = []
+    def __init__(self) -> None:
+        self.operations: List[Callable[..., Any]] = []
 
-    def __call__(self, sequence):
-        return reduce(lambda x, operation: operation(x), self.operations, sequence)
+    def __call__(self, sequence: Sequence[Any]) -> Any:
+        return reduce(lambda x, operation: operation(x), self.operations, sequence)  # type: ignore
 
-    def apply(self, function):
-        def operation(sequence):
+    def apply(self, function: Callable[[Any], Any]) -> "Pipeline":
```

Comment:
Do these functions actually get called with anything? Or should it be a `Record` like filter?

Reply:
`.apply()` is called with functions later in this file, I'll see if there is something in common between them so I can make the signature more specific.
```python
pipeline = (
    Pipeline()
	...
    .apply(sort_group_contents)
    .apply(sort_rule_groups)
)
```